### PR TITLE
Major refactor of attribute handling, test and diagnostic improvements

### PR DESCRIPTION
**Pull request by jhwgh1968**
In doing this refactoring, I also:
* Fixed a bug causing nested items to be generated when they should not have been.
* Fixed a small oversight in the Strings package revealed by the fix above.
* Added a test case for handling "overload style" renames, while investigating the previous item.
* Added a better diagnostic for the problem mentioned in [this comment](https://github.com/schungx/rhai/issues/38#issuecomment-678758747).

### CODE REVIEW

**Comment by jhwgh1968** [comment]
> Added a better diagnostic for the problem mentioned in [this comment](https://github.com/schungx/rhai/issues/38#issuecomment-678758747).

For the record, I managed to reproduce the issue you were having (I think) with this test code:

```rust
#[export_fn(return_raw)]
pub fn test_fn(input: Point) -> bool /* wrong return type */ {
    input.x > input.y
}

fn main() {
    let n = Point {
        x: 0.0,
        y: 10.0,
    };
    if test_fn(n) {
        println!("yes");
    } else {
        println!("no");
    }
}
```

After adding a diagnostic test and fixing the issue as best I could, this is now the output on stable Rust:

```
error[E0308]: mismatched types
  --> $DIR/export_fn_raw_return.rs:10:8
   |
9  | #[export_fn(return_raw)]
   | ------------------------ expected `std::result::Result<rhai::any::Dynamic, std::boxed::Box<rhai::result::EvalAltResult>>` because of return type
10 | pub fn test_fn(input: Point) -> bool {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `std::result::Result`, found `bool`
   |
   = note: expected enum `std::result::Result<rhai::any::Dynamic, std::boxed::Box<rhai::result::EvalAltResult>>`
              found type `bool`

error[E0308]: mismatched types
  --> $DIR/export_fn_raw_return.rs:10:33
   |
9  | #[export_fn(return_raw)]
   | ------------------------ expected `std::result::Result<rhai::any::Dynamic, std::boxed::Box<rhai::result::EvalAltResult>>` because of return type
10 | pub fn test_fn(input: Point) -> bool {
   |                                 ^^^^ expected enum `std::result::Result`, found `bool`
   |
   = note: expected enum `std::result::Result<rhai::any::Dynamic, std::boxed::Box<rhai::result::EvalAltResult>>`
              found type `bool`
```

It's a bit wordy, but hopefully should be clear.

**Comment by schungx** [comment]
> After adding a diagnostic test and fixing the issue as best I could, this is now the output on stable Rust:

I think another case would be returning inconsistent types from two if branches:

```rust
if input.x > input.y {
    Ok(true.into())
} else {
    Ok(input.x.into())
}
```

**Comment by schungx** [comment]
OK, I tested the error diagnostics.

`#[export_fn]` successfully puts out an error for `return_raw` which does not return `Result`.
`#[export_fn]` functions have good error diagnostics.

`#[export_module]` functions do not highlight `return_raw` errors, not do they highlight interior errors.

Therefore, it seems like we're still missing error position info for modules.

**Comment by jhwgh1968** [comment]
> Therefore, it seems like we're still missing error position info for modules.

Hmm. That is a bit of a surprise, but then again, I didn't write a separate test for it.

I will fix this.
